File: solutions/103.binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if(root == null) return res;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        List<Integer> curRes = null;
        TreeNode curNode = null;
        boolean reverse = false;
        while(!queue.isEmpty()) {
            curRes = new ArrayList<>();
            for(int size = queue.size(); size > 0; size--) {
                curNode = queue.remove();
                if(reverse) curRes.add(0, curNode.val);
                else curRes.add(curNode.val);
                if(curNode.left != null) queue.offer(curNode.left);
                if(curNode.right != null) queue.offer(curNode.right);
            }
            res.add(curRes);
            reverse = !reverse;
        }
        return res;
    }
}

File: solutions/103.binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        dfs(res, root, 0);
        return res;
    }
private:
    void dfs(vector<vector<int>>& res, TreeNode* root, int level) {
        if(!root) return;
        if(res.size() == level) 
            res.push_back(vector<int>());
        if(level % 2 == 0) res[level].push_back(root->val);
        else res[level].insert(res[level].begin(), root->val);
        dfs(res, root->left, level + 1);
        dfs(res, root->right, level + 1);
    }
};